azureDatabaseUtils/src/main/java/bio/terra/workspace/azureDatabaseUtils/database/DatabaseService.java

```diff
@@ -80,4 +90,157 @@ public void restoreNamespaceRoleAccess(String namespaceRole) {
+      String dbName,
+      String dbHost,
+      String dbPort,
+      String adminUser,
+      String blobFileName,
+      String blobContainerName,
+      String blobContainerUrlAuthenticated) {
+
+    // Grant the database role (dbName) to the landing zone identity (adminUser).
+    // In theory, we should be revoking this role after the operation is complete.
+    // We are choosing to *not* revoke this role for now, because:
+    // (1) we could run into concurrency issues if multiple users attempt to clone the same
+    // workspace at once;
+    // (2) the workspace identity can grant itself access at any time, so revoking the role
+    // doesn't protect us.
+    databaseDao.grantRole(adminUser, dbName);
+
+    List<String> commandList = generateCommandList("pg_dump", dbName, dbHost, dbPort, adminUser);
+    Map<String, String> envVars = null;
+    try {
+      envVars = Map.of("PGPASSWORD", determinePassword());
+    } catch (PSQLException e) {
+      logger.error(e.getMessage());
+    }
+
+    logger.info("running DatabaseService.pgDump: {}", String.join(" ", commandList));
+    logger.info("blobContainerName: {}", blobContainerName);
+
+    LocalProcessLauncher localProcessLauncher = new LocalProcessLauncher();
+    localProcessLauncher.launchProcess(commandList, envVars);
+
+    storage.streamOutputToBlobStorage(
+        localProcessLauncher.getInputStream(),
+        blobFileName,
+        blobContainerName,
+        blobContainerUrlAuthenticated);
+
+    checkForError(localProcessLauncher);
+  }
+
+  public void pgRestore(
+      String dbName,
+      String dbHost,
+      String dbPort,
+      String adminUser,
+      String blobFileName,
+      String blobContainerName,
+      String blobContainerUrlAuthenticated) {
+
+    boolean doCleanup = false;
+
+    // Grant the database role (dbName) to the workspace identity (adminUser).
+    // In theory, we should be revoking this role after the operation is complete.
+    // We are choosing to *not* revoke this role for now, because:
+    // (1) we could run into concurrency issues if multiple users attempt to clone the same
+    // workspace at once;
+    // (2) the workspace identity can grant itself access at any time, so revoking the role
+    // doesn't protect us.
+    try {
+      databaseDao.grantRole(adminUser, dbName);
+
+      List<String> commandList = generateCommandList("psql", dbName, dbHost, dbPort, adminUser);
+      Map<String, String> envVars = null;
+      try {
+        envVars = Map.of("PGPASSWORD", determinePassword());
+      } catch (PSQLException e) {
+        logger.error(e.getMessage());
+      }
+      LocalProcessLauncher localProcessLauncher = new LocalProcessLauncher();
+      localProcessLauncher.launchProcess(commandList, envVars);
+
+      logger.info("running DatabaseService.pgRestore: {}", String.join(" ", commandList));
+      logger.info("blobContainerName: {}", blobContainerName);
+
+      storage.streamInputFromBlobStorage(
```

Comment:
in the WDS case, we cant just restore the [schema ](https://github.com/DataBiosphere/terra-workspace-data-service/blob/main/service/src/main/java/org/databiosphere/workspacedataservice/service/BackupRestoreService.java#L192)from the pg_dump and leave it as is -- we actually need to change the schema definition from old workspace id to new workspace id (since the pg_dump we take comes from the old workspace). In this code, I dont see either of those values (workspaceId or sourceWorkspaceId) passed into these functions -- would you expect that this needs to happen outside of this code? My preference would be to keep it here since if an error does happen, this is contained in the restore code where we wont leave the user with a non usable wds db schema. 

not sure if cbas needs similar alteration since as far as I know it doesnt have a concept of instances?

Reply:
Any transformations that need to happen after the pg_dump are app specific. So anything WDS needs to do will have to be up to WDS and we will need to figure out a way for WDS to know that it is starting up from a clone. CBAS wants to clear out any workflow runs from the source workspace. I think their plan is to delete any record older than the database creation time which will come from WSM and passed in by Leo.

Reply:
> not sure if cbas needs similar alteration since as far as I know it doesnt have a concept of instances?

that's right - we won't need to do this kind of thing for cbas. Doug's comment about our plan is correct

Reply:
originally I didn't like the idea of having a database with a schema that was tied to another workspace Id, but I chatted with David more about this and mostly resolved all my concerns (around WDS somehow becoming unusable due to this or us having unnecessary data in the db/etc)